### ISSUE TITLE
Biome Stone and Y Min

### ISSUE THREAD

**Issue by C1ffisme** [minetest/minetest#8233]
##### Issue type
 - Feature request
 
##### Minetest version
```
5.0.0
```

##### Summary

When registering a biome, you can set a stone node which will generate below that biome all the way until `y_min`. This is very useful for desert stone and biomes that fit the Vanilla Minetest game. However, I have a different goal in mind.

I'd like to generate a biome that will only appear if the surface is from `y_min` to `y_max`, but I want a certain type of stone to appear all the way to the bottom of the world. This means the player has to explore the world horizontally to find biomes with certain stones, ores, etc. The problem is that while I could just set `y_min` to -32000, the biome generates ugly ocean/riverbeds:

![screenshot_20190214_190049](https://user-images.githubusercontent.com/9893266/52827886-32f54400-308c-11e9-887e-69028aff1445.png)

What I'd like to suggest is that instead of generating stone all the way to `y_min`, a biome's stone will generate all the way down to wherever the coder specifies using a different paramter, and `y_min`/`y_max` will only be used to specify where the biome will spawn on the surface.

**Comment by paramat** [comment]
> When registering a biome, you can set a stone node which will generate below that biome all the way until `y_min`. This is very useful for desert stone and biomes that fit the Vanilla Minetest game

It's actually a general and flexible system useful for biomes of many games, it's not vanilla-centric or MTG-centric.
> The problem is that while I could just set `y_min` to -32000, the biome generates ugly ocean/riverbeds:

Ugly in what way? Is the underwater surface node in the foreground your stone node and is that the problem? You can add surface nodes to underwater biomes to hide your chosen stone node. MTG does this, whatever the stone node is it is hidden below sand nodes.
Any mapgen with sloping rivers (Valleys) has a definable riverbed node with definable depth, that will also hide stone.
I suspect this can be solved with careful use of biomes. Can you link to the biome registrations? I can help out with these.
(You seem to have surface snow nodes underwater too).
> I'd like to generate a biome that will only appear if the surface is from `y_min` to `y_max`

> What I'd like to suggest is that instead of generating stone all the way to `y_min`, a biome's stone will generate all the way down to wherever the coder specifies using a different paramter, and `y_min`/`y_max` will only be used to specify where the biome will spawn on the surface.

This can't happen by default obviously as it would mess up all biomes.
It's possible you don't need this feature and the problem is Biome API usage. However, if you do need it, it is a specialised behaviour so is not justified. The Biome API can only contain extremely useful and heavily used functionality.

**Comment by C1ffisme** [comment]
> Ugly in what way? Is the underwater surface node in the foreground your stone node and is that the problem?

That is not my stone node, it's my `node_top`. The problem is that the top node generates underwater.

> You can add surface nodes to underwater biomes to hide your chosen stone node. MTG does this, whatever the stone node is it is hidden below sand nodes.

That's the problem, I'm not using an underwater biome. I would like to switch to using one, but in order to do so I would have to set the `y_min` of the biome to 0 or higher, and stone would only generate to that level. There are only two solutions to my problem:

A - Add a parameter for what node is placed under the ocean, similar to what we have for rivers
B - Add a parameter for how deep stone generates.

In other words, I'd like to be able to do this:

![biome api change](https://user-images.githubusercontent.com/9893266/52831520-8cfd0600-309a-11e9-8d42-914b5fe47a82.png)

But at the moment, I can only do this:

![biome api change current](https://user-images.githubusercontent.com/9893266/52831528-94241400-309a-11e9-82ed-ad1e9be80250.png)

The idea is to add more types of stone that generate in different biomes rather than just the bland palette of "mapgen_stone", and to avoid any stone becoming the "Main Stone type"

**Comment by paramat** [comment]
> That is not my stone node, it's my `node_top`. The problem is that the top node generates underwater.

Yes looking at your screenshot the ugliness is caused by not splitting your biomes into 'above water' and 'below water' pairs as MTG does.
> That's the problem, I'm not using an underwater biome. I would like to switch to using one, but in order to do so I would have to set the `y_min` of the biome to 0 or higher, and stone would only generate to that level.

I can't see the problem. You can split the biome into 2 biomes, one above water, one below water, both having the desired biome stone, and the 'below water' biome having suitable underwater surface nodes, just as MTG does. So the desired biome stone wouldn't stop at water level, you can stop it at any depth using 'y_min' and placing a 3rd biome below it.

Your second diagram is not possible, the beach stone will extend across the whole diagram (in 5.0.0-dev anyway, maybe 0.4.16/17 too, a while back behaviour was slightly different).
> A - Add a parameter for what node is placed under the ocean, similar to what we have for rivers

That was your initial request, it's actually not practically possible.
When an underground mapchunk is generated, to know whether to place an 'under land' or 'under sea' stone node it needs to know the structure of terrain at water level. So every underground mapchunk would have to also generate the water level mapchunk above it. Mapgen would become twice as intensive.
So this is a 'can't add' issue.
> B - Add a parameter for how deep stone generates.

This is already possible.
Start with a set of biomes distributed horizontally according to heat and humidity noises.
Vertically split each biome into 2 biomes: 'above water' and 'below water' versions, each with suitable surface nodes.
For the biome where you want to specify how deep the biome stone generates, vertically split that into 3 biomes: 'above water', 'below water and above biome stone y_min', 'below biome stone y_min'.
> The idea is to add more types of stone that generate in different biomes

Yes it's already possible and games are doing this.

This is probably a Biome API usage issue and you probably don't need the feature, however, it's not possible anyway. I'll close the issue for now but feel free to continue the discussion and i'll try to help you achieve what you want and we can clear up any misunderstandings if there are any.

**Comment by C1ffisme** [comment]
> For the biome where you want to specify how deep the biome stone generates, vertically split that into 3 biomes: 'above water', 'below water and above biome stone y_min', 'below biome stone y_min'.

This doesn't work though. I still don't think I've clarified my point enough, so let me explain exactly what I'm trying to do:

Unlike Minetest Game, I'm trying to avoid a single "stone" node that generates under the entire world, instead I plan on using different kinds of stone. Using the ore generation system is not an option, since my goal is to avoid having a main type of stone, and using the ore generation system would create ugly blotches of stone rather than large areas.

Instead, I'm adding multiple forests/deserts/jungles with different kinds of stone underneath each biome. (Granite, Gniess, Slate, Limestone, etc.) I want the layer of Granite/Gniess/Whatever Biome-Specific Rock to be placed underneath the biome down to stone_min_y.

Right now I've been able to place stone underneath the biome by extending the biome's min_y all the way down to the bottom of the world. The only problem is that this means I can't specify at what point on the surface a biome should spawn.

Creating a third biome "below biome stone_min_y" is not possible, since biomes are not generated in 3d, they are generated in a 2d grid. This other biome below stone_min_y would just generate as another biome wherever the ocean floor dropped down below where I want the min_y for my biome to be.

What I'm essentially suggesting is that biome-specific stone be spawned as far underneath a biome as I need it to go, regardless of where the biome can spawn on the surface.

> Your second diagram is not possible, the beach stone will extend across the whole diagram (in 5.0.0-dev anyway, maybe 0.4.16/17 too, a while back behaviour was slightly different).

The second diagram is possible, it's what is currently in the game. The biome-specific stone for the beach will extend all the way to the beach's y-min, which is probably the bottom of the world so that the beach biome will spawn even in the deepest oceans. (You seem to have confused the beach stone for `mapgen_stone`. That is not the case, I wanted to show that the ocean biome's stone would spawn all the way from just below the surface to y_min)

The Forest, on the other hand, cannot spawn its own biome-specific stone below y_min. If I set the y_min lower, the forest would spawn where that beach is. If I add another biome to spawn in that area, it would spawn next to the beach, and not under the forest. This means that if I dig underneath a Forest, I will always reach a layer of `mapgen_stone` when I hit the y-min.

**Comment by Ezhh** [comment]
It's already possible to avoid mapgen_stone by having a specific stone for each biome. I appreciate this isn't quite what you are looking for, but it creates some interesting and diverse underground. It would also be quite weird to have different stone under oceans. The stone from the main biome would usually extend beneath it.

A big problem here seems to be how you'd detect where exactly an ocean is, since they could have different heat/humidity values, and though you can select biome height, shore biomes would generally need to include all nodes at a specific height for a given heat/humidity, regardless of whether the land height falls within that height range. 

Could you share what biome definitions you'd use to get your second diagram? Code could show where you are currently far more easily than words.

**Comment by C1ffisme** [comment]
> It's already possible to avoid mapgen_stone by having a specific stone for each biome.

That's what I'm trying to do, the problem is that the specific stone only appears down to y_min.

> Could you share what biome definitions you'd use to get your second diagram? Code could show where you are currently far more easily than words.

The second diagram would look something like what we already have in minetest_game:

```
default.register_alias("mapgen_stone","mapgen:unwanted_default_stone")

minetest.register_biome({
  name = "Forest"
  ...
  node_stone = "forest:stone",
  y_min = 5,
})

minetest.register_biome({
  name = "Beach"
  ...
  node_stone = "beach:stone"
  y_max = 5
  y_min = -32000
})
```

I feel like I've caused too much confusion by focusing too much on the ocean biome problem, so here's a flowchart so I can describe the larger problem:

![my problem](https://user-images.githubusercontent.com/9893266/52874735-07b83680-3118-11e9-8b06-fbcba626ee71.png)

**Comment by Ezhh** [comment]
If I'm understanding this, and it's possible I'm not, you want to set the biome for any x,z coordinate based on the y value of land level at that point?

The current biome system works from heat and humidity values instead, so you're never going to achieve this with less than a whole new system. With that said... maybe I can help you a bit anyway.

Have you looked into stacking biomes so that you layer them? This can give some really nice effects - for example, we could do something like this:

![biomes ideas](https://user-images.githubusercontent.com/7035183/52890019-8219a280-317a-11e9-9d20-02b24c0528a1.png)

In this example, I'd tend to think of forest and desert as the main biome types, but I'd be registering multiple actual biomes for each to have height specific sub divisions within them. This way you can have different stone types in the below ground layers, and different top/filler nodes on the above ground ones. 

Because shore biomes are usually limited to a few nodes above water level, you can easily insert shore biomes as follows:

![biomes ideas2](https://user-images.githubusercontent.com/7035183/52890219-72e72480-317b-11e9-8698-5a9770b89cbd.png)

The sand (or other node as desired) that you will see on the shore should be a top node, so it won't show as a band right through the map when you dig down from higher areas.

To my thinking, this approach has an advantage, because:
- Each stack of biomes can share the same or similar stone types and broad theme.
- You can design neighbouring biomes to have nice logical progressions (I'd suggest thinking about heat/humidity values here - a rain forest shouldn't border a glacier and so on).

This makes it easier to get logical progressions that make sense both vertically and horizontally, while still leaving you with a lot of control over content.

**Comment by paramat** [comment]
> Unlike Minetest Game, I'm trying to avoid a single "stone" node that generates under the entire world,

MTG doesn't have to be like that, each biome can have a special stone running to world base.
> Right now I've been able to place stone underneath the biome by extending the biome's min_y all the way down to the bottom of the world.

> The only problem is that this means I can't specify at what point on the surface a biome should spawn.

The second observation is unrelated to the first statement. The only way to horizontally distribute biomes is by heat and humidity noises, there is no other way, it's practically impossible to place a certain biome stone only under ocean areas or only under land areas.
This is true whether you extend a biome to world base or not.
> Creating a third biome "below biome stone_min_y" is not possible, since biomes are not generated in 3d,

They are, using y_max and y_min.
First you create a system of biomes horizontally distributed according to heat and humidity noises.
Next you split each biome vertically into multiple copies, each with the same heat and humidity points. You can stack as many copies as you want to place a particular biome stone anywhere you want.
> What I'm essentially suggesting is that biome-specific stone be spawned as far underneath a biome as I need it to go, regardless of where the biome can spawn on the surface.

Yes already possible by stacking.
> The second diagram is possible, it's what is currently in the game. The biome-specific stone for the beach will extend all the way to the beach's y-min, which is probably the bottom of the world so that the beach biome will spawn even in the deepest oceans.

No it isn't possible. A while back the biome stone would correspond to the surface biome, but only to the mapchunk base, not to world base. Now that has changed in 5.0.0 (or earlier), a biome's y_min is strictly enforced even for stone.
> The Forest, on the other hand, cannot spawn its own biome-specific stone below y_min. If I set the y_min lower, the forest would spawn where that beach is. If I add another biome to spawn in that area, it would spawn next to the beach, and not under the forest. This means that if I dig underneath a Forest, I will always reach a layer of `mapgen_stone` when I hit the y-min.

That's because you are expecting underground biomes to depend on whether there is ocean above, they can't, there is only an illusion of that happening by that being the case down to mapchunk base, in older versions of MT.

Horizontally, biomes can only be located according to heat and humidity noises, not by anything else like being under ocean or land areas.

Are you actually using 'heat point' and 'humidity point' in your biome definitions? If you are only vertically stacking biomes using y limits then no wonder you're having issues.
Please can you post your definitions in full or link to them?

In your flowchart statements at bottom and left are incorrect.

The biome system can be tricky to work with so i can understand the confusion.
Ezhh's posts are correct and good advice.

**Comment by C1ffisme** [comment]
Well, paramat was right about the beaches, but this code:

```minetest.register_biome({
	name = "gniess_plains",
	y_min = 2,
	y_max = 100,
	node_stone = "mapgen:gniess",
	node_filler = "mapgen:loam",
	depth_filler = 3,
	node_top = "mapgen:loam_grass",
	depth_top = 1,
	node_riverbed = "mapgen:flint_sand",
	heat_point = 70,
	humidity = 60,
})

minetest.register_biome({
	name = "gniess_beach",
	y_min = -2000,
	y_max = 1,
	node_stone = "mapgen:marble",
	node_filler = "mapgen:loam",
	depth_filler = 3,
	node_top = "mapgen:flint_sand",
	depth_top = 1,
	node_riverbed = "mapgen:flint_sand",
	heat_point = 70,
	humidity = 80,
})

minetest.register_biome({
	name = "clay_underground",
	y_min = -2000,
	y_max = 1,
	node_stone = "mapgen:clay",
	node_filler = "mapgen:clay",
	depth_filler = 3,
	node_top = "mapgen:clay",
	depth_top = 1,
	node_riverbed = "mapgen:clay",
	heat_point = 70,
	humidity = 60, -- SAME AS GNIESS PLAINS, This is the "Underbiome" of the Gniess Plains
})
```

Generated no clay biome, for some reason it only generated a beach biome.

[This is my current biome code.](https://gist.github.com/C1ffisme/af39e9eb979cb7d7fad129917a7e0256)

**Comment by paramat** [comment]
Heat and humidity noises are 2D noises in X and Z.

Note that freely varying y limits and heat/humidity points can cause results that are tricky to control. As a result i learnt to first divide horizontally using heat/humidity points, then divide each heat/humidity point vertically using y limits, as in MTG. Good luck.

**Comment by C1ffisme** [comment]
Thank you paramat, I'm sorry for my uninformed ramble.

**Comment by paramat** [comment]
Hey no problem, biomes are tricky.
